core/trino-spi/src/main/java/io/trino/spi/exchange/Exchange.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.exchange;
+
+import io.airlift.slice.Slice;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.Closeable;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+@ThreadSafe
+public interface Exchange
+        extends Closeable
+{
+    /**
+     * Registers a new sink
+     *
+     * @param taskPartitionId uniquely identifies a dataset written to a sink
+     * @return {@link ExchangeSinkHandle} associated with the <code>taskPartitionId</code>.
+     * Must be passed to {@link #instantiateSink(ExchangeSinkHandle, int)} along with an attempt id to create a sink instance
+     */
+    ExchangeSinkHandle addSink(int taskPartitionId);
+
+    /**
+     * Called when no more sinks will be added with {@link #addSink(int)}.
+     * New sink instances for an existing sink may still be added with {@link #instantiateSink(ExchangeSinkHandle, int)}.
+     */
+    void noMoreSinks();
+
+    /**
+     * Registers a sink instance for an attempt.
+     * <p>
+     * Attempts are expected to produce the same data.
+     * <p>
+     * The implementation must ensure the data written by unsuccessful attempts is safely discarded.
```

Comment:
You may want to note when it is supposed to happen (on `close()` I believe)

Reply:
I would prefer to leave it for an implementer to decide where and how to discard. The important part is that the data from failed / duplicated attempts shouldn't be visible.